### ISSUE TITLE
如果实现动态UserAgent和动态秘钥

### ISSUE THREAD

**Issue by andrew0213** [code4craft/webmagic#633]
1.想要实现，每次请求时，都对Header进行一次处理，把UserAgent设置为我UserAgent数组的一个随机值，
2.使用的是一个动态代理，每次请求都要重新生成一个秘钥值,在header中传过去

在scrapy中，可以在middlewares中处理，比较优雅，但是现在我没找到一个合适的位置去做处理，求解惑。

**Comment by andrew0213** [comment]
最好是，也能有一个地方，达到request到downloader的钩子程序的效果,能根据各自需求，做一些加工.

**Comment by luchatex** [comment]
每次new Request时候可以处理,或者poll出来的request,替换成你自己新的header

**Comment by andrew0213** [comment]
嗯，已经找到了。但现在遇到一个新的问题，我要捕获请求失败的情况，比如504码,做记录，spider中有个onDownloadSuccess方法做了处理，但方法是private，没法重写。请问我还有其他方式吗？

**Comment by code4craft** [comment]
同意 @luchatex 的方案，可以用代理模式在普通的scheduler外面封装一层。scrapy的middlewares也挺不错的，类似于Java的filter。

WebMagic之前版本的设计初衷是模块化和完全的自由定制，所以希望通过替换组件的形式来完成自定义，组件本身其实扩展性并不强(Scheduler、Downloader)。

这样的问题就是基于WebMagic的fork很多，然后基本上都各有不同，很难统一化。随着WebMagic的受众越来越多，浅度定制的需求也慢慢有了必要。后面除了功能的完善，工程上的易用性也会是主要目标。

**Comment by luchatex** [comment]
@andrew0213 状态码是可以自定义的,除了可以在Site里使用 setAcceptStatCode(acceptStatCode)来自定义成功码以外,还可以自定义一个downloader重写 protected void onSuccess(Request request) 这个方法记录你需要的请求参数.

但我是在自己定义的Processor里统一解决的,无论是抓取成功还是失败,还有进行下一次爬取的情况,其实webmagic是很灵活的,譬如我需要再每次访问的时候发送一个请求token才可以获得数据.
那么我就重写Scheduler的Poll方法,然后把里面的参数修改下,返回自己定制的Request即可.

但请注意一并修改protected void pushWhenNoDuplicate(Request request, Task task) 方法,这个方法就是在你请求失败(可能是各种原因,譬如代理失败)后,再次放回队列的.
